modules/drivers/redshift/src/metabase/driver/redshift.clj

```diff
@@ -251,3 +251,10 @@
                               :filter_values       (field->parameter-value query)})
        " */ "
        (qputil/default-query->remark query)))
+
+(defmethod sql-jdbc.sync/have-any-schema-privileges? :redshift [_ ^Connection conn ^String table-schema]
+  (with-open [stmt ^PreparedStatement (prepare-statement conn "SELECT HAS_SCHEMA_PRIVILEGE(?, 'USAGE');")]
+    (.setString stmt 1 table-schema)
+    (with-open [rs ^ResultSet (.executeQuery stmt)]
```

Comment:
1.  Just add tag information to `prepare-statement` rather than having to tag it every time it's used.

   ```clj
   (defn- prepare-statement ^PreparedStatement [^Connection conn sql]
     (.prepareStatement conn sql
                        ResultSet/TYPE_FORWARD_ONLY
                        ResultSet/CONCUR_READ_ONLY
                        ResultSet/CLOSE_CURSORS_AT_COMMIT))
   ``` 

2.  (For future reference) generally you'd add the tag to the symbol e.g. `^PreparedStatement stmt (...)` instead of on the fn call form e.g. `stmt ^PreparedStatement (...)` -- although the latter does work. It looks unidiomatic tho

3. You don't need to tag Java methods like `.executeQuery`, Java is statically typed so it already knows `Statement.executeQuery` returns a `ResultSet` without you telling it

4.  You could combine this into one `with-open` statement if you did 

   ```clj
   (doto (prepare-statement ...) 
     (.setString 1 table-schema))
   ```

Reply:
If you follow my suggestion and made a `remove-schemas-without-privileges` or `syncable-schemas` function you could optimize this a bit further. You can create the prepared statement just once and reuse it (just set a new string and fetch `with-open` a new result set for each schema). Since a big database might have a few thousand schemas we should try to optimize this a bit if we can

Reply:
This is done via a stateful transducer now in the Redshift version.